### PR TITLE
Feature/node pool

### PR DESCRIPTION
**Pull request by demonwy**

### CODE REVIEW

File: alicloud/resource_alicloud_cs_kubernetes_node_pool_test.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+						"tags.0.key":            "test-tag-key",
+						"tags.0.value":          "test-tag-value",
+						"tags.1.key":            "new-tag-key",
+						"tags.1.value":          "new-tag-value",
+						"data_disks.#":          "2",
+						"data_disks.0.size":     "100",
+						"data_disks.0.category": "cloud_ssd",
+						"data_disks.1.size":     "200",
+						"data_disks.1.category": "cloud_efficiency",
+					}),
```
Comment (zql-tqy) [inline alicloud/resource_alicloud_cs_kubernetes_node_pool_test.go:90]: 添加 resource_group_id 然后进行变更的 step，资源组的问题比较多

File: alicloud/provider.go [hunk 2]
```diff
@@ -418,6 +418,7 @@ func Provider() terraform.ResourceProvider {
 			"alicloud_cs_managed_kubernetes":               resourceAlicloudCSManagedKubernetes(),
 			"alicloud_cs_serverless_kubernetes":            resourceAlicloudCSServerlessKubernetes(),
 			"alicloud_cs_kubernetes_autoscaler":            resourceAlicloudCSKubernetesAutoscaler(),
+			"alicloud_cs_kubernetes_node_pool":             resourceAlicloudCSKubernetesNodePool(),
```
Comment (zql-tqy) [inline alicloud/provider.go:421]: 这个也需要在 erb 文档中进行添加

File: alicloud/resource_alicloud_cs_kubernetes_node_pool.go [hunk 3]
```diff
@@ -0,0 +1,715 @@
...
+
+	nodePool, ok := raw.(*cs.CreateNodePoolResponse)
+	if ok != true {
+		return WrapErrorf(err, DefaultErrorMsg, "alicloud_cs_kubernetes_node_pool", "ParseKubernetesNodePoolResponse", raw)
+	}
+
+	d.SetId(nodePool.NodePoolID)
+
+	// reset interval to 10s
+	stateConf := BuildStateConf([]string{"initial", "scaling"}, []string{"active"}, d.Timeout(schema.TimeoutCreate), 10*time.Second, csService.CsKubernetesNodePoolStateRefreshFunc(d.Get("cluster_id").(string), d.Id(), []string{"deleting", "failed"}))
```
[the remark this answers is not in the document]
Comment (demonwy) [reply alicloud/resource_alicloud_cs_kubernetes_node_pool.go:255]: 如果count=0的情况下，这个时间差不多；否则需要分钟级.

File: alicloud/resource_alicloud_cs_kubernetes_node_pool.go [hunk 4]
```diff
@@ -0,0 +1,715 @@
...
+			return WrapErrorf(fmt.Errorf("node_count can not be less than before"), "scaleOutFailed %d:%d", newValue, oldValue)
+		}
+		args.Count = int64(newValue) - int64(oldValue)
+		d.SetPartial("node_count")
+	}
+	args.NodePoolInfo.Name = d.Get("name").(string)
+	if d.HasChange("name") && !d.IsNewResource() {
+		update = true
+		args.NodePoolInfo.Name = d.Get("name").(string)
+		d.SetPartial("name")
```
[the remark this answers is not in the document]
Comment (demonwy) [reply alicloud/resource_alicloud_cs_kubernetes_node_pool.go:300]: 这里面的d.SetPartial("name")都去掉是吗?

File: alicloud/resource_alicloud_cs_kubernetes_node_pool.go [hunk 5]
```diff
@@ -0,0 +1,715 @@
...
+		}
+
+		stateConf := BuildStateConf([]string{"scaling"}, []string{"active"}, d.Timeout(schema.TimeoutUpdate), 10*time.Second, csService.CsKubernetesNodePoolStateRefreshFunc(d.Get("cluster_id").(string), d.Id(), []string{"deleting", "failed"}))
+
+		if _, err := stateConf.WaitForState(); err != nil {
+			return WrapErrorf(err, IdMsg, d.Id())
+		}
+	}
+
+	update = false
```
[the remark this answers is not in the document]
Comment (demonwy) [reply alicloud/resource_alicloud_cs_kubernetes_node_pool.go:458]: 好的

File: alicloud/resource_alicloud_cs_kubernetes_node_pool.go [hunk 6]
```diff
@@ -0,0 +1,715 @@
...
+	update := false
+	invoker := NewInvoker()
+	args := &cs.UpdateNodePoolRequest{
+		RegionId:         common.Region(client.RegionId),
+		NodePoolInfo:     cs.NodePoolInfo{},
+		ScalingGroup:     cs.ScalingGroup{},
+		KubernetesConfig: cs.KubernetesConfig{},
+		AutoScaling:      cs.AutoScaling{},
+	}
+	if d.HasChange("node_count") && !d.IsNewResource() {
```
[the remark this answers is not in the document]
Comment (demonwy) [reply alicloud/resource_alicloud_cs_kubernetes_node_pool.go:277]: 好的，这里适合后者； 创建完直接走update

File: website/docs/r/cs_kubernetes_node_pool.html.markdown [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+This resource will help you to manager node pool in Kubernetes Cluster. 
+
+-> **NOTE:** Available in 1.94.0+.
+
+## Example Usage
+
+node_pool in Kubernetes Cluster
+
+```
```
[the remark this answers is not in the document]
Comment (demonwy) [reply website/docs/r/cs_kubernetes_node_pool.html.markdown:20]: ok

File: website/docs/r/cs_kubernetes_node_pool.html.markdown [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+sidebar_current: "docs-alicloud-resource-cs-kubernetes-node-pool"
+description: |-
+  Provides a Alicloud resource to manage container kubernetes node pool.
+---
+
+# alicloud\_cs\_kubernetes\_node\_pool
+
+This resource will help you to manager node pool in Kubernetes Cluster. 
+
+-> **NOTE:** Available in 1.94.0+.
```
[the remark this answers is not in the document]
Comment (demonwy) [reply website/docs/r/cs_kubernetes_node_pool.html.markdown:14]: 好的

File: alicloud/resource_alicloud_cs_kubernetes_node_pool.go [hunk 9]
```diff
@@ -0,0 +1,715 @@
...
+		return WrapError(err)
+	}
+
+	d.Set("node_count", object.TotalNodes)
+	d.Set("cluster_id", d.Get("cluster_id").(string))
+	d.Set("name", object.Name)
+	d.Set("vpc_id", object.VpcId)
+	d.Set("vswitch_ids", object.VswitchIds)
+	d.Set("instance_types", object.InstanceTypes)
+	//d.Set("password", object.LoginPassword)
```
[the remark this answers is not in the document]
Comment (demonwy) [reply alicloud/resource_alicloud_cs_kubernetes_node_pool.go:481]: 可以删掉

File: alicloud/resource_alicloud_cs_kubernetes_node_pool_test.go (hunk 1 above)
Comment (demonwy) [reply alicloud/resource_alicloud_cs_kubernetes_node_pool_test.go:90]: 什么意思？

File: alicloud/provider.go (hunk 2 above)
Comment (demonwy) [reply alicloud/provider.go:421]: 具体在哪个文档？

File: alicloud/resource_alicloud_cs_kubernetes_node_pool.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (demonwy) [reply alicloud/resource_alicloud_cs_kubernetes_node_pool.go:255]: 这里改成30s

**Comment by huyngopt1994** [comment]
Error: Unsupported argument

  on ../../../modules/alicloud/managed_kubernetes/main.tf line 24, in resource "alicloud_cs_kubernetes_node_pool" "custom":
  24:   labels = []

An argument named "labels" is not expected here. Did you mean to define a
block of type "labels"?
Hi team When I use this resouce, the labels and taints doesn't work ?

**Comment by xiaozhu36** [comment]
@huyngopt1994 The labels is a list and you should split them to multi block, like:
```
    lalels = {
      key = "xxxx"
      value = "ddd"
   }
    lalels = {
      key = "xxxx"
      value = "ddd"
   }
    lalels = {
      key = "xxxx"
      value = "ddd"
   }
```

**Comment by huyngopt1994** [comment]
@xiaozhu36 I figured out the error, labels is block not attribute, I should set like this 

```
resource "alicloud_cs_kubernetes_node_pool" "default" {
  for_each       = var.node_pools
  cluster_id     = alicloud_cs_managed_kubernetes.k8s.id
  name           = each.key
  node_count     = each.value.node_count
  vswitch_ids    = var.worker_vswitch_ids
  instance_types = each.value.instance_types
  tags           = each.value.tags
  labels {
    key = "kafka-od"
    value ="true"
  }

  depends_on     = [alicloud_cs_managed_kubernetes.k8s]
}```

**Pull request merged**